README.md
# Prometheus CUPS Exporter

This exporter provides prometheus metrics for cups by utilizing pycups.

By default it will register on Port 9329 as registered in the [Prometheus Wiki](https://github.com/prometheus/prometheus/wiki/Default-port-allocations)

```text
usage: cups_exporter.py [-h] [--cups-host CUPS_HOST] [--cups-port CUPS_PORT]
                        [--cups-user CUPS_USER] [--listen-port LISTEN_PORT]

optional arguments:
  -h, --help            show this help message and exit
  --cups-host CUPS_HOST
                        The cups host to connect to (default: localhost)
  --cups-port CUPS_PORT
                        The cups port to use (default: 631)
  --cups-user CUPS_USER
                        The user to connect with (default: default)
  --listen-port LISTEN_PORT
                        The port the exporter will listen on (default: 9329)
```

The metrics exported are:

```text
# HELP cups_print_jobs_active Number of current print jobs
# TYPE cups_print_jobs_active gauge
cups_print_jobs_active 0.0
# HELP cups_print_jobs_total Total number of print jobs
# TYPE cups_print_jobs_total counter
cups_print_jobs_total 1000.0
# HELP cups_printers Number of printers
# TYPE cups_printers gauge
cups_printers 20.0
# HELP cups_printer_status_info Status about printer alerts
# TYPE cups_printer_status_info gauge
cups_printer_status_info{model="Good Printers Inc. Printer",printer="My-Printer",status="happy"} 1.0
# HELP cups_up CUPS status
# TYPE cups_up gauge
cups_up 1.0
# HELP cups_scrape_duration_seconds Amount of time each scrape takes
# TYPE cups_scrape_duration_seconds gauge
cups_scrape_duration_seconds 0.04441666603088379
```


Changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

- [#5](https://github.com/ThoreKr/cups_exporter/pull/5) Add a Dockerfile (and add automatic builds on DockerHub)
- [#7](https://github.com/ThoreKr/cups_exporter/issues/6) Fix a bug where incomplete responses were returned if the list of print jobs was empty.
- [#7](https://github.com/ThoreKr/cups_exporter/issues/7) Do not keep a persistent connection to cups to survive cups restarts.

## [1.0.0] - 2019-09-25

### Added

- Gather on scrape time
- Added `cups_scrape_duration_seconds` metric
- Add Pipfile for dependency management

### Changed

- Renamed some metrics to adhere to naming conventions
  - `cups_printer_status` -> `cups_printer_status_info`
  - `cups_print_jobs` -> `cups_print_jobs_total`


cups-exporter.py
#!/usr/bin/env python3
import cups
import json
import time
from prometheus_client import start_http_server, Gauge
import argparse

parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument("--cups-host", help="The cups host to connect to", default="localhost")
parser.add_argument("--cups-port", type=int, help="The cups port to use", default=631)
parser.add_argument("--cups-user", help="The user to connect with", default="default")
parser.add_argument("--listen-port", type=int, help="The port the exporter will listen on", default=9329)
args = parser.parse_args()


# Define Metrics
printJobsNum = Gauge('cups_print_jobs', 'Number of current print jobs')
printersNum = Gauge('cups_printers', 'Number of printers')
printersStatus = Gauge('cups_printer_status', 'Status about printer alerts', ['printer', 'status'])
dhcpUp = Gauge('cups_up', 'CUPS up')


#@printersNum
def getPrinterData(conn):
  printers = conn.getPrinters()
  printersNum.set(len(printers))
  return printers

#@printJobsNum
def getJobData(conn):
  jobs = conn.getJobs()
  printJobsNum.set(len(jobs))

#@printersStatus
def getPrinterStatus(printers):
  for key, value in printers.items():
    if value['printer-state-reasons'][0] != 'none':
      printersStatus.labels(printer=value['printer-make-and-model'], status=value['printer-state-reasons'][0]).set(0)
    else:
      printersStatus.labels(printer=value['printer-make-and-model'], status='happy').set(1)


if __name__ == '__main__':
  # Start up the server to expose the metrics.
  start_http_server(args.listen_port)

  cups.setServer(args.cups_host)
  cups.setPort(args.cups_port)
  cups.setUser(args.cups_user)

  while True:
    try:
      conn = cups.Connection()
      printers = getPrinterData(conn)
      getJobData(conn)
      getPrinterStatus(printers)
      dhcpUp.set(1)
    except Exception as e:
      dhcpUp.set(0)
      print(e)

    time.sleep(5)


cups_exporter.py
#!/usr/bin/env python3
"""Export CUPS metrics to prometheus
"""

import argparse
import time

import cups
from prometheus_client import start_http_server
from prometheus_client.core import (REGISTRY, CounterMetricFamily,
                                    GaugeMetricFamily)

parser = argparse.ArgumentParser(
    formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument(
    "--cups-host", help="The cups host to connect to", default="localhost")
parser.add_argument("--cups-port", type=int,
                    help="The cups port to use", default=631)
parser.add_argument(
    "--cups-user", help="The user to connect with", default="default")
parser.add_argument("--listen-port", type=int,
                    help="The port the exporter will listen on", default=9329)
args = parser.parse_args()


class CUPSCollector:
    """CUPSCollector collects status data about
    currently configured printers in cups
    """

    def __init__(self, host, port, user):
        """Set cups connection parameters

        Arguments:
            host {str} -- Cups Hostname
            port {int} -- Cups Port
            user {str} -- Cups username
        """
        self._prometheus_metrics = {}

        self.host = host
        self.port = port
        self.user = user


    def collect(self):
        """Collects the metrics from cups
        """
        start = time.time()

        self._setup_empty_prometheus_metrics()

        cups.setServer(self.host)
        cups.setPort(self.port)
        cups.setUser(self.user)

        try:
            conn = cups.Connection()

            printers = conn.getPrinters()
            self._prometheus_metrics['printersNum'].add_metric(
                [],
                len(printers))

            self._getPrinterStatus(printers)
            self._getJobData(conn)

            self._prometheus_metrics['cupsUp'].add_metric([], 1)
        except Exception as e:
            self._prometheus_metrics['cupsUp'].add_metric([], 0)
            print(e)

        duration = time.time() - start
        self._prometheus_metrics['scrape_duration_seconds'].add_metric(
            [], duration)

        for metric in self._prometheus_metrics.values():
            yield metric

    def _setup_empty_prometheus_metrics(self):
        """
        The metrics we want to export.
        """
        self._prometheus_metrics = {
            'printJobsNum':
                GaugeMetricFamily('cups_print_jobs_active',
                                  'Number of current print jobs'),
            'printJobsTotal':
                CounterMetricFamily('cups_print_jobs_total',
                                    'Total number of print jobs'),
            'printersNum':
                GaugeMetricFamily('cups_printers',
                                  'Number of printers'),
            'printerStatus':
                GaugeMetricFamily('cups_printer_status_info',
                                  'Status about printer alerts',
                                  labels=['printer', 'model', 'status']),
            'cupsUp':
                GaugeMetricFamily('cups_up',
                                  'CUPS status'),
            'scrape_duration_seconds':
                GaugeMetricFamily('cups_scrape_duration_seconds',
                                  'Amount of time each scrape takes',
                                  labels=[])
        }

    def _getJobData(self, conn):
        """Collects data about cups
        """
        jobs = conn.getJobs(which_jobs="all")
        if len(jobs) == 0:
            self._prometheus_metrics['printJobsTotal'].add_metric([], 0)
        else:
            lastjobID = list(jobs.keys())[-1]
            self._prometheus_metrics['printJobsTotal'].add_metric(
                [], lastjobID)

        jobs = conn.getJobs()
        if len(jobs) == 0:
            self._prometheus_metrics['printJobsNum'].add_metric([], 0)
        else:
            self._prometheus_metrics['printJobsNum'].add_metric(
                [],
                len(jobs))

    def _getPrinterStatus(self, printers):
        """Gathers printer status data

        Arguments:
            printers {dict} -- dict of printers
        """
        for key, value in printers.items():
            if value['printer-state-reasons'][0] != 'none':
                self._prometheus_metrics['printerStatus'].add_metric([
                    key,
                    value['printer-make-and-model'],
                    value['printer-state-reasons'][0]],
                    0)
            else:
                self._prometheus_metrics['printerStatus'].add_metric([
                    key,
                    value['printer-make-and-model'],
                    'happy'],
                    1)


if __name__ == '__main__':
    # Start up the server to expose the metrics.
    REGISTRY.register(CUPSCollector(
        args.cups_host,
        args.cups_port,
        args.cups_user))
    start_http_server(args.listen_port)
    while True:
        time.sleep(1)
